### PR TITLE
[DataPipe] Extend FileLister to support load multiple directories

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#72260 [DataPipe] Extend FileLister to support load multiple directories**

Change `FileLister` to support single directory, multiple directories and files. This is needed for TorchAudio revamping especially for `OnDiskCacheHolder`

Differential Revision: [D33979744](https://our.internmc.facebook.com/intern/diff/D33979744)

### CODE REVIEW

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33979744).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33979744).

File: torch/utils/data/datapipes/iter/__init__.py [hunk 1]
```diff
@@ -1,3 +1,6 @@
+from torch.utils.data.datapipes.iter.utils import (
+    IterableWrapperIterDataPipe as IterableWrapper,
+)
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/__init__.py:3]: Re-order the import to prevent circular import.

File: torch/utils/data/datapipes/iter/filelister.py [hunk 2]
```diff
@@ -1,22 +1,24 @@
...
 class FileListerIterDataPipe(IterDataPipe[str]):
     r""" :class:`FileListerIterDataPipe`
 
     Iterable DataPipe to load file pathname(s) (path + filename), yield pathname from given disk root dir.
 
     Args:
-        root: Root directory
+        root: Root directory or a sequence of root directories
         mask: Unix style filter string or string list for filtering file name(s)
         abspath: Whether to return relative pathname or absolute pathname
```
Comment (NivekT) [inline torch/utils/data/datapipes/iter/filelister.py:15]: Can you update the `args` description to include the arguments `recursive`, `abspath`, `non_deterministic`? Maybe length too?

Comment (NivekT) [reply torch/utils/data/datapipes/iter/filelister.py:15]: `non_deterministic: if False, the results yielded from each root directory will be sorted`

Comment (NivekT) [reply torch/utils/data/datapipes/iter/filelister.py:15]: Update `mask` to `masks`

**Review by NivekT — APPROVED** [review]
LGTM with some note. I don't think this is BC-breaking so no need for extra label. Can we add this to 1.11 or does it need to wait?

**Comment by ejguan** [comment]
> LGTM with some note. I don't think this is BC-breaking so no need for extra label. Can we add this to 1.11 or does it need to wait?

We may not be able to bring it into 1.11 because branch cut has done and this PR doesn't meet any Cherry-Pick Criteria IMO. We can still let users to use nightly.

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33979744).

**Comment by ejguan** [comment]
@NivekT Second thought. I will try to convince release team to make this PR merged into 1.11. Otherwise, we have to create another `FileLister` in TorchData to override it.
